lazy_loader/__init__.py

```diff
@@ -257,15 +271,72 @@ def attach_stub(package_name: str, filename: str):
         incorrectly (e.g. if it contains an relative import from outside of the module)
     """
     stubfile = (
-        filename if filename.endswith("i") else f"{os.path.splitext(filename)[0]}.pyi"
+        filename if filename.endswith("i")
+        else f"{os.path.splitext(filename)[0]}.pyi"
     )
 
     if not os.path.exists(stubfile):
-        raise ValueError(f"Cannot load imports from non-existent stub {stubfile!r}")
+        raise ValueError(
+            f"Cannot load imports from non-existent stub {stubfile!r}",
+        )
 
     with open(stubfile) as f:
         stub_node = ast.parse(f.read())
 
     visitor = _StubVisitor()
     visitor.visit(stub_node)
     return attach(package_name, visitor._submodules, visitor._submod_attrs)
+
+
+class LazyFinder(MetaPathFinder):
+
+    def find_spec(
+        self,
+        fullname: str,
+        path: Sequence[str] | None,
+        target: types.ModuleType | None = ...,
+        / ,
+    ) -> ModuleSpec | None:
+        if not inside_context_manager.get():
+            # We are not in context manager, delegate to normal import
+            return None
+
+        if searching_spec.get():
+            # We are searching for the loader, so we should continue the search
+            return None
+
+        searching_spec.set(True)
+        spec = importlib.util.find_spec(fullname)
+        searching_spec.set(False)
+
+        if spec is None:
+            raise ModuleNotFoundError(f"No module named '{fullname}'")
+
+        spec.loader = importlib.util.LazyLoader(spec.loader)
```

Comment:
Since the lazy loader produces a delayed evaluation such that any import errors will show up later than expected, and since the user of the `lazy_imports` context may not realize what is happening under the hood, it might make sense to subclass the LazyLoader and do something like
```
class MyLazyLoader(LazyLoader):
    def exec_module(self, module):
        try:
            super().exec_module(module)
        except BaseException as e:
            raise ImportError(f"Lazy load of module {module} failed") from e

Reply:
I in principle agree with improving the user reporting of errors, so this sounds fine to me.